README.md
![icon]

Tasu
====

![Tests status](https://github.com/yentsun/tasu/workflows/Tests/badge.svg?branch=master)
[![Coverage Status](https://coveralls.io/repos/github/yentsun/tasu/badge.svg?branch=master)](https://coveralls.io/github/yentsun/tasu?branch=master)
[![Known Vulnerabilities](https://snyk.io/test/github/yentsun/tasu/badge.svg?targetFile=package.json)](https://snyk.io/test/github/yentsun/tasu?targetFile=package.json)

An `async/await` wrapper over [node-nats](https://github.com/nats-io/node-nats),
designed to easily integrate with your microservice code. Taşuu (ташуу)
is 'transport' in Kyrgyz.


Installation
------------

```
npm i tasu
```

Usage
-----

Create an instance of the transport:

```js
import Tasuu from 'tasu';

async function main()  {
    ...
    const tasu = new Tasuu({ group: 'some-service', ...});
    await tasu.connected();
}

```


Publish a request and get a response via `tasu.request()` on one end:

```js
const { bar } = await tasu.request('foo', { rg: 1 });
```
_Note: this method uses `requestOne` inside, no need to worry about max  
responses_ 


Subscribe and respond to a request on the other:

```js
tasu.listen('foo', async ({ arg }, respond) => {
    const bar = await someDatabaseCall(arg);
    return { bar };
});
```


_Note: A listener is automatically added to queue group `foo.listeners`;
errors of the handling function are caught and sent back as error response_


Publish an event on one end:

```js
tasu.publish('some.package.sent', {...});
```

Subscribe and process as worker queue on the other:

```js
 tasu.process('*.package.sent', (pack, subject) => {
    console.log(subject, pack);
});
```

`listen`, `subscribe` and `process` methods return an integer
subscription ID (SID) which can be used to unsubscribe from a subject:

```js
const sid = tasu.process('*.package.sent', (pack, subject) => {
    console.log(subject, pack);
});

...

tasu.unsubscribe(sid);
```
The above technique is useful for websocket connections.


Close NATS connection (if needed):

```js
tasu.close();
```


Configuration
-------------

You can set the following config options while creating a new Tasu
instance:

- `url` - NATS connection urls, default: `nats://localhost:4222`
- `group` - group name this transport will be assigned to, default: `default`
- `requestTimeout` - how long to wait for response on NATS requests,
  default: `10000`
- `level` - set log level for provided logger. Default is `debug`
- `logger` - define custom logger. Should basically have `info`,
  `debug`, `error` methods.


API
---

- **connected()** - returns a promise resolved on successful connection to
  NATS server. You basically always want this resolved before proceeding
  with Tasu.

- **publish(subject, message)** - *optimistically* publish a `message`
  to a `subject`. TODO: make this return a promise too.

- **listen(subject, async (message)=>{...})** - subscribes to a `subject`
  and respond via handler function. Handler function must be defined via
  `async`, it gets `message` object as the only argument and *must
  return something* as successful response. Errors are caught and sent
  back as error response. Returns subscription id.

- **subscribe(subject, (message, replyTo, subject)=>{...})** - subscribes
  to a `subject` and process *messages* with handler function. All
  memebers of the *group* will get subject messages. Returns subscription
  id.

- **subOnce(subject, (message, subject)=>{...})** - same as `subscribe()`
  but unsubscribes immediately after receiving a message.

- **process(subject, (message, subject)=>{...})** - subscribes to a
  `subject` as a queue worker and processes *messages* via handler
  function. Messages will be distributed randomly among free group
  members. Returns subscription id.

- **request(subject\[, message\])** - performs a request and returns a
  response promise.

- **unsubscribe(sid)** - unsubscribes from subscription identified by
  `sid` (returned by `subscribe`, `subOnce`, `process`, `listen`)

- **close()** - closes connection to NATS. Always try to exit gracefully,
  otherwise the connection will persist until next heartbeat.


Credits
-------

Icons by [icons8](https://icons8.com)

[icon]: icons8-hub-64.png


CHANGELOG.md
[5.0.0] - 2020-12-13
--------------------
- [x] Move to ESM modules #28
- [x] Drop `chai` dependency


[4.3.1] - 2020-10-19
--------------------
- [x] Update dependencies


[4.3.0] - 2019-07-06
--------------------
- [x] Drop `lodash.merge` dependency
- [x] Bump `nats` version to 1.3.0
- [x] Bump Go version to 1.11 (for test NATS process)


[4.2.1] - 2018-07-17
--------------------
- [x] Increase Go version to 1.9
- [x] Fix vulnerabilities


[4.1.1] - 2018-04-15
--------------------
- [x] Update `trid` version


[4.1.0] - 2018-04-05
--------------------
- [x] Throw errors instead of emitting
- [x] Drop `EventEmitter` extension
- [x] Fix documentation #7
- [x] Remove formatter from logger options
- [x] Allow `request(subject)` without messages


[4.0.1] - 2018-04-02
--------------------
- [x] Fix error detail tests
- [x] Drop `moment` dependency


[4.0.0] - 2018-04-02
--------------------
- [x] Remove `respond` callback from `listen`
- [x] Move to `trid` for ids
- [x] Formattable logger


[3.3.0] - 2018-03-30
--------------------
- [x] Add custom error class #8


[3.2.2] - 2018-03-30
--------------------
- [x] Change `nats` to `tasu` @ logger


[3.2.1] - 2018-03-22
--------------------
- [x] Add `engines` to npm package


[3.2.0] - 2018-02-22
--------------------
- [x] Cover code to 100% (well, 99% actually) #6


[3.1.3] - 2018-02-22
--------------------
- [x] Add more keywords to `package.json`
- [x] Add coveralls with nyc #5


[3.1.0] - 2018-02-19
--------------------
- [x] Add 'async style' connected state handling
- [x] Update dependencies


[3.0.0] - 2017-12-19
--------------------
- [x] Move requests to promise-based scheme fix #1


[2.1.2] - 2017-10-13
--------------------
- [x] Remove 'debug' level from stderr stream @ logger


[2.1.1] - 2017-09-30
--------------------
- [x] ADDED deeper password screening


[2.1.0] - 2017-09-28
--------------------
- [x] ADDED one-time subscription


[2.0.2] - 2017-09-28
--------------------
- [x] FIXED some typos


[2.0.0] - 2017-09-15
--------------------
- [x] ADDED ids to request/response
- [x] ADDED `connected` event handling
- [x] CHANGE sinon mocks to live NATS connection @ tests
- [x] REMOVED js standard


[1.6.0] - 2017-08-02
--------------------
- [x] ADDED password screening


[1.5.4] - 2017-07-03
--------------------
- [x] ADDED more info @ error response logging


[1.5.3] - 2017-05-21
--------------------
- [x] ADDED `unsubscribe` method
- [x] ADDED more info @ error response logging


[1.5.2] - 2017-05-17
--------------------
- [x] ADDED error.stack to error response log entry


[1.5.1] - 2017-05-13
--------------------
- [x] ADDED option to override default logger


[1.5.0] - 2017-05-10
--------------------
- [x] ADDED `group` parameter to queue subscriptions


[1.4.0] - 2017-04-15
--------------------
- [x] CHANGED `response` behavior to always return JSON array with error as first element and message as the second


[1.3.4] - 2017-04-14
--------------------
- [x] ADDED `close` method which closes underlying NATS connection


[1.3.3] - 2017-04-04
--------------------
- [x] FIXED empty response handling


[1.3.2] - 2017-04-04
--------------------
- [x] FIXED options loading


[1.3.1] - 2017-03-17
--------------------
- [x] FIXED error serialization


[1.3.0] - 2017-03-17
--------------------
- [x] ADDED `respond` function as part of returned params from `listen`


[1.2.0] - 2017-03-17
--------------------
- [x] ADDED `listen` method to subscribe to point-to-point requests
- [x] Standard JS polish


[1.1.0] - 2017-03-15
--------------------
- [x] ADDED readme
- [x] ADDED `respond` method
- [x] ADDED `process` method to subscribe to a queue group
- [x] REMOVE `options` argument from `subscribe`


[1.0.0] - 2017-03-12
--------------------
- [x] CHANGED request response to return unwrapped object


[0.2.2] - 2017-03-08
--------------------
- [x] CHANGED request logging from verbs to arrows


[0.2.1] - 2017-03-03
--------------------
- [x] CHANGED default log level to `debug`


[0.2.0] - 2017-03-01
--------------------
- [x] ADDED `error.detail` (pg) support


[0.1.2] - 2017-02-28
--------------------
- [x] ADDED default options


[0.1.1] - 2017-02-28
--------------------
- [x] CHANGED initial response timeout to 1000


[0.1.0] - 2017-02-28
--------------------
- [x] ADDED `request` method
- [x] ADDED `publish` method
- [x] ADDED `subscribe` method


test.js
import assert  from 'assert';
import Tasuu from './';
import sinon from 'sinon';


describe('tasu: empty options', () => {

    const tasu = new Tasuu();

    before(async () => {
        await tasu.connected();
    });

    after(async () => {
        tasu.close();
    });

    it('runs ok', (done) => {
        assert.ok(tasu.id);
        assert.strictEqual(tasu.group, 'default');
        done();
    });
});


describe('tasu: options set', () => {

    const tasu = new Tasuu({  // a nats server should be running
        url: 'nats://localhost:4222',
        group: 'tests',
        requestTimeout: 100
    });

    before(async () => {
        await tasu.connected();
    });

    describe('request', () => {

        before((done) => {
            tasu.listen('request.ok', (message) => {
                return message;
            });
            tasu.listen('request.password', (message) => {
                return message;
            });
            tasu.listen('request.error', () => {
                throw Error('service error');
            });
            tasu.listen('request.empty', () => {
                return null;
            });
            tasu.listen('request.error.detail', () => {
                const error = Error('service error');
                error.detail = 'error detail';
                delete error.message;
                throw error;
            });
            done();
        });

        it('performs a request and returns successful result', async () => {
            const {foo} = await tasu.request('request.ok', {foo: 'bar'});
            assert.strictEqual(foo, 'bar');
        });

        it('performs a request and screens password field in logs', async () => {
            const response = await tasu.request('request.password', {
                password: 'should not see this',
                data: {password: 'should not see this also'}
                });
            assert.strictEqual(response.password, 'should not see this');
            assert.strictEqual(response.data.password, 'should not see this also');
        });

        it('performs a request and returns error', async () => {
            try {
                await tasu.request('request.error', {foo: 'bar'});
            } catch (error) {
                assert.strictEqual(error.message, 'service error');
                assert.ok(error.requestId);
            }
        });

        it('performs a request and returns an empty response', async () => {
            const response = await tasu.request('request.empty', { foo: 'bar' });
            assert.strictEqual(response, null);
        });

        it('performs a request and returns error from error.detail', async () => {
            try {
                await tasu.request('request.error.detail', { foo: 'bar' });
            } catch (error) {
                assert.strictEqual(error.message, 'error detail');
                assert.ok(error.requestId);
            }
        });

        it('performs a request and returns timeout error', async () => {
            try {
                await tasu.request('request.timeout', { foo: 'bar' });
            } catch (error) {
                assert.strictEqual(error.message, 'response timeout');
                assert.ok(error.requestId);
            }
        }).timeout(110);
    });

    describe('subscribe', () => {

        it('performs a response-subscription to a subject', (done) => {
            tasu.subscribe('broadcast', (message, replyTo, subject) => {
                assert.strictEqual(message.foo, 'bar');
                assert.ok(! replyTo);
                assert.strictEqual(subject, 'broadcast');
                done()
            });
            tasu.publish('broadcast', {foo: 'bar'});
        })
    });

    describe('listen', () => {

        function timeout(ms) {
            return new Promise(resolve => setTimeout(resolve, ms));
        }

        it('responds with object', async () => {

            tasu.listen('request.listen', async (message) => {
                assert.strictEqual(message.foo, 'bar');
                await timeout(5);
                return {bar: 'foo'};
            });
            const {bar} = await tasu.request('request.listen', {foo: 'bar'});
            assert.strictEqual(bar, 'foo');
        });

        it('responds with number', async () => {

            tasu.listen('request.listen.number', async (message) => {
                await timeout(5);
                return 5;
            });
            const bar = await tasu.request('request.listen.number');
            assert.strictEqual(bar, 5);
        });

    });

    describe('process', () => {

        it('processes a queue message as group member', (done) => {
            tasu.process('process', (message, subject) => {
                assert.strictEqual(message.foo, 'bar');
                assert.strictEqual(subject, 'process');
                done()
            });
            tasu.publish('process', {foo: 'bar'});
        })
    });

    describe('publish', () => {

        it('publishes a message to a subject', (done) => {
            tasu.process('publish', (message) => {
                assert.strictEqual(message.message, 'yay!');
                done();
            });
            tasu.publish('publish', {message: 'yay!'});
        });
    });

    describe('unsubscribe', () => {

        it('unsubscribes from NATS subject', (done) => {
            const sid = tasu.subscribe('unsubscribe', ()=>{});
            tasu._nats.once('unsubscribe', (subId, subject) => {
                assert.strictEqual(subId, sid);
                assert.strictEqual(subject, 'unsubscribe');
                done();
            });
            tasu.unsubscribe(sid);
        })
    });

    describe('subOnce', () => {

        it('subscribes, gets one message and unsubscribes', (done) => {
            const sid = tasu.subOnce('subOnce', (message) => {
                assert.strictEqual(message.foo, 'once');
            });
            tasu.publish('subOnce', {foo: 'once'});
            tasu._nats.once('unsubscribe', (subId, subject) => {
                assert.strictEqual(subId, sid);
                assert.strictEqual(subject, 'subOnce');
                done();
            });
        });
    });

    describe('nats error event', () => {

        it('trows error', (done) => {
            try {
                const error = new Error('synthetic error');
                tasu._nats.emit('error', error);
            } catch (error) {
                assert.strictEqual(error.message, 'synthetic error');
                done()
            }

        });

        it('exits on connection error', (done) => {
            const sandbox = sinon.sandbox.create({useFakeTimers : true});
            const exitStub  = sandbox.stub(process, 'exit');
            const error = new Error('synthetic connection error');
            error.code = 'CONN_ERR';
            tasu._nats.emit('error', error);
            sinon.assert.calledOnce(exitStub);
            sandbox.restore();
            done();
        })
    });

    describe('connection events', () => {

        it('handles disconnect', (done) => {
            tasu._nats.emit('disconnect');
            assert.strictEqual(tasu._state, 'disconnected');
            done();
        });

        it('handles reconnecting', (done) => {
            tasu._nats.emit('reconnecting');
            assert.strictEqual(tasu._state, 'reconnecting');
            done();
        });

        it('handles reconnect', (done) => {
            tasu._nats.emit('reconnect');
            assert.strictEqual(tasu._state, 'connected');
            done();
        });
    });

    describe('close', () => {

        it('closes underlying connection with NATS', (done) => {
            tasu.close();
            assert(tasu._nats.closed);
            done();
        })
    });

});


lib/logger.js
import winston from 'winston';
import screenPassword from './screenPassword';


export default function ({ group, level='debug' }) {

    const formatter = function (options) {
        return [options.timestamp(), 'tasu.' + group, options.level.toUpperCase(), options.message || '', options.meta && Object.keys(options.meta).length ? JSON.stringify(options.meta) : ''].join(' ')
    };

    return new (winston.Logger)({
        rewriters: [screenPassword],
        transports: [
            new (winston.transports.Console)({
                level,
                timestamp: function () {
                    const now = new Date();
                    return now.toISOString()
                },
                formatter,
                stderrLevels: ['error']
            })
        ]
    });
};


index.js
import nats from 'nats';
import TRID from 'trid';
import Logger from './lib/logger';
import RequestError from './lib/RequestError';


export default class {

    constructor(options) {

        const defaults = {
            url: 'nats://localhost:4222',
            requestTimeout: 10000,
            group: 'default',
        };
        this._options = {...defaults, ...options};
        const {group, level} = this._options;
        this._logger = this._options.logger || Logger({group, level});
        this._nats = nats.connect(options);
        this._trid = new TRID({prefix: this._options.group});
        this.id = this._trid.base();
        this.group = this._options.group;

        this.readyPromise = new Promise(resolve => {

            this._nats.on('connect', () => {
                this._state = 'connected';
                this._logger.info('connected to NATS server:', this._nats.currentServer.url.host);
                this._logger.info('id:', this.id);
                this._logger.info('group:', this._options.group);
                resolve()
            });
        });

        this._nats.on('error', (error) => {

            this._logger.error(`${error.message}${error.code ? ' (code: '+error.code+')' : ''}`);
            if (error.code === 'CONN_ERR') {
                process.exit(1);
            }
            else
                throw error;
        });

        this._nats.on('disconnect', () => {

            if (this._state === 'connected') {
                this._logger.error('nats disconnected');
            }
            this._state = 'disconnected';
        });

        this._nats.on('reconnecting', () => {

            this._state = 'reconnecting';
            this._logger.info('nats reconnecting');
        });

        this._nats.on('reconnect', () => {

            this._state = 'connected';
            this._logger.info('nats reconnected');
        });
    }


    connected() {

        return this.readyPromise;
    }


    publish(subject, message) {

        this._logger.debug('publishing to', subject, message);
        this._nats.publish(subject, JSON.stringify(message), () => {
            this._logger.debug('message published', subject, message)
        })
    };


    // returned by `listen`, not to be used directly

    _respond(error, replyTo, response) {

        if (error) {
            this._logger.debug('sending error response to', replyTo, error);
            this._nats.publish(replyTo, JSON.stringify([{message: error.message || error.detail, stack: error.stack}]), () => {
                this._logger.debug('error response sent to', replyTo)
            })
        } else {
            this._nats.publish(replyTo, JSON.stringify([null, response]), () => {
            })
        }
    };


    // subscribe to point-to-point requests

    listen(subject, done) {

        const group = subject + '.listeners';
        this._logger.debug('subscribing to requests', subject, 'as member of', group);
        return this._nats.subscribe(subject, {queue: group}, async (message, reply, subject) => {
            try {
                const result = await done(JSON.parse(message),subject);
                this._respond(null, reply, result);
            } catch (error) {
                this._respond(error, reply);
            }
        });
    };


    // subscribe to broadcasts

    subscribe(subject, done) {

        this._logger.debug('subscribing to broadcasts', subject);
        return this._nats.subscribe(subject, (message, replyTo, subject) => {
            this._logger.debug('got broadcast', subject, message);
            done(JSON.parse(message), replyTo, subject);
        })
    };


    // subscribe as queue worker

    process(subject, done) {

        const group = subject + '.workers.' + this._options.group;
        this._logger.debug('subscribing to process', subject, 'queue as member of', group);
        return this._nats.subscribe(subject, {queue: group}, (message, reply, subject) => {
            this._logger.debug('processing', subject, message);
            done(JSON.parse(message), subject);
        })
    };


    // request one response

    request(subject, message={}) {

        const meta = JSON.parse(JSON.stringify(message));  // important to clone here, as we are rewriting meta
        const id = this._trid.seq();
        this._logger.debug('[>>', id, '>>]', subject, meta);
        return new Promise((resolve, reject) => {
            this._nats.requestOne(subject, JSON.stringify(message), null, this._options.requestTimeout, (response) => {
                if (response.code && response.code === nats.REQ_TIMEOUT) {
                    this._logger.error('[!!', id, '!!] timeout', subject, message);
                    reject(new RequestError('response timeout', id));
                } else {
                    const [error, res] = JSON.parse(response);
                    if (error) {
                        this._logger.error('[!!', id, '!!] ', error.message, error.stack);
                        reject(new RequestError(error.message, id));
                    } else {
                        const meta = JSON.parse(JSON.stringify(res));
                        this._logger.debug('[<<', id, '<<]', subject, meta);
                        resolve(res);
                    }
                }
            })
        });
    };


    // unsubscribe from subscription by id

    unsubscribe(sid) {

        this._logger.debug(`unsubscribing from ${sid}`);
        this._nats.unsubscribe(sid);
    };


    // subscribe, receive, unsubscribe

    subOnce(subject, done) {

        this._logger.debug(`subscribing once to ${subject}...`);
        const sid = this._nats.subscribe(subject, (message, replyTo, subject) => {
            this._logger.debug('got message', subject, message, '- unsubscribing from', sid);
            this._nats.unsubscribe(sid);
            done(JSON.parse(message), subject);
        });
        return sid;
    }


    // close underlying connection with NATS

    close() {
        this._logger.info('closing connection with NATS:', this._nats.currentServer.url.host);
        this._nats.close();
    }
};